docs/sql/query_syntax/with.md

```diff
@@ -5,20 +5,166 @@ selected: Documentation/SQL/Query Syntax/With
+  SELECT id, name, [name] AS path
+  FROM tag
+  WHERE subclassof IS NULL
+UNION ALL
+  SELECT tag.id, tag.name, list_prepend(tag.name, tag_hierarchy.path)
+  FROM tag, tag_hierarchy
+  WHERE tag.subclassof = tag_hierarchy.id
+)
+SELECT path
+FROM tag_hierarchy
+WHERE source = 'Oasis';
+```
+```
+┌───────────────────────────┐
+│           path            │
+├───────────────────────────┤
+│ [Oasis, Rock, Music, Art] │
+└───────────────────────────┘
+```
 
+#### Graph traversal
+
+The `WITH RECURSIVE` clause can be used to express graph traversal on arbitrary graphs. However, if the graph has cycles, the query must perform cycle detection to prevent infinite loops.
+
+Take the following undirected social graph:
+
+```sql
+CREATE TABLE knows(person1id int, person2id int);
+INSERT INTO knows VALUES (1, 2), (1, 4), (2, 3), (2, 4), (3, 4), (5, 6);
+INSERT INTO knows SELECT person2id, person1id FROM knows;
+```
+
+Note that there is a cycle e.g. between nodes 1, 2, and 4. To detect cycles, the query stores the path in a [list](/docs/sql/data_types/list) and, before adding a new edge, checks whether its endpoint has been visited before.
+
+The following query returns all paths from person 1:
+
+```sql
+WITH RECURSIVE paths(startPerson, endPerson, path) AS (
+   SELECT -- define the path as the first edge of the traversal
+        person1id AS startPerson,
+        person2id AS endPerson,
+        [person1id, person2id]::bigint[] AS path
+     FROM knows
+   UNION ALL
+   SELECT -- concatenate new edge to the path
+        paths.startPerson AS startPerson,
+        person2id AS endPerson,
+        array_append(path, person2id) AS path
+     FROM paths
+     JOIN knows ON paths.endPerson = knows.person1id
+    WHERE knows.person2id != ALL(paths.path) -- detect cycles
+)
+SELECT startPerson, endPerson, path
+FROM paths
+WHERE startPerson = 1;
+```
+```
+┌─────────────┬───────────┬──────────────┐
+│ startPerson │ endPerson │     path     │
+├─────────────┼───────────┼──────────────┤
+│ 1           │ 2         │ [1, 2]       │
+│ 1           │ 4         │ [1, 4]       │
+│ 1           │ 3         │ [1, 4, 3]    │
+│ 1           │ 4         │ [1, 2, 4]    │
+│ 1           │ 2         │ [1, 4, 2]    │
+│ 1           │ 3         │ [1, 2, 3]    │
+│ 1           │ 2         │ [1, 4, 3, 2] │
+│ 1           │ 3         │ [1, 2, 4, 3] │
+│ 1           │ 4         │ [1, 2, 3, 4] │
+│ 1           │ 3         │ [1, 4, 2, 3] │
+└─────────────┴───────────┴──────────────┘
+```
+
+`WITH RECURSIVE` can be used to find all unweighted shortest paths between two nodes. The following query returns all unweighted shortest paths between person 1 and person 3:
+
+```sql
+WITH RECURSIVE paths(startPerson, endPerson, path, endPersonReached) AS (
+   SELECT person1id AS startPerson,
+          person2id AS endPerson,
+          [person1id, person2id]::bigint[] AS path,
+          max(CASE WHEN person2id = 3 THEN true ELSE false END) OVER () AS endPersonReached
```

Comment:
No ordering is specified here, so I worry that the data could get reordered if `paths` gets large. It doesn't matter for this small example, but if someone were to use it for a much larger problem, could this be an issue?

Reply:
Good point... I'm not sure.

My thinking is that `max()` is associative, so `endPersonReached` is `true` if *any* of the `person2id`s in the subquery results is 3 (indicating that the graph traversal has reached the end person in that step).

Reply:
> My thinking is that `max()` is associative, so `endPersonReached` is `true` if _any_ of the `person2id`s in the subquery results is 3 (indicating that the graph traversal has reached the end person in that step).

It is, but the default frame bounds goes from the start of the partition to the current row, so it will only be set for rows where there is a 3 there or before, which might not be correct if the data was reordered somehow.

Reply:
Right. So one way to keep them in sync is to sort it with total ordering. For example, the first subquery would become:
```sql
SELECT person1id AS startPerson,
        person2id AS endPerson,
        [person1id, person2id]::bigint[] AS path,
        max(CASE WHEN person2id = 3 THEN true ELSE false END) OVER (ORDER BY person1id, person2id) AS endPersonReached
    FROM knows
WHERE person1id = 1
ORDER BY person1id, person2id
```
This may be a performance hit – I have to microbenchmark it – but if it helps avoid a (very infrequent) Heisenbug, I'm all for it.

Reply:
I don't think the ordering is needed, just expand the frame to be the entire table? `ORDER (UNBOUNDED PRECEDING TO UNBOUNDED FOLLOWING)` would just set the flag to be true if any path in the table has reached the end.

Reply:
Excellent, thanks! Have not used window functions much, so this helps a lot. I sent a follow-up PR: #161